### PR TITLE
Add argument randomization

### PR DESCRIPTION
**Pull request by bernhl**
This PR adds support for randomizing the command line arguments of the js engines.
The command line of fuzzilli now takes an additional optional argument `--argumentRandomization=` which is either `none` (default), `initial` (the instances to randomize the arguments once when starting fuzzilli) or `periodic` (periodic re-randomization of command line arguments).
Currently, randomization is supported for v8, spidermonkey and jsc only. Adding support for randomizing command line args of additional engines should be relatively straightforwards.

Any feedback on the PR is welcome, its plenty of changes.

### CODE REVIEW

**Review by saelo — COMMENTED** [review]
Cool idea, thanks a lot!

So on a high level I'm wondering if we can do something simpler. Wouldn't it be enough to serialize the program arguments to a string already in the Profiles, and then keeping just the string (as `let additionalArguments: String` or so) associated with the program? The arguments could then just be randomly regenerated every hour or so for example. In REPRL, I guess we'd need to split up the string again, and restart the engine if the string has changed, but we wouldn't need to encode engine-specific arguments in protobufs and pass them around. It'd all be contained to the (engine-specific) Profiles. Would that work as well?

**Comment by bernhl** [comment]
Reducing the engine-specific parts in the serialized programs/protobufs is certainly feasible but I'm not sure whether we can avoid changing the protobuf entirely. The workers need to communicate the arguments to the master instance which executes the test case again to decide whether it is flaky or not. Adding an array of strings to the protobuf (containing the additional args) should be sufficient and the serialized format wouldn't be engine specific. However, the command line args make no sense for other engines and trying to import them will result in sub-optimal results.

Regarding the configuration options for randomization: right now the possible values are `none/initial/periodic`. I could change it to `none`, `periodic=timeInSeconds` so the interval you proposed as one hour would be configurable; the `initial` setting would be mood as one can just set a really large value.

**Comment by saelo** [comment]
So I'd like to try and reduce this down to a minimal set of changes (I'm in general currently trying to somewhat aggressively trim down Fuzzilli since there's too much logic that's essentially unmaintained...), so sorry for the additional effort :/ and please let me know if you disagree!

I'm wondering if something as simple as this is good enough:
- Only regenerate random arguments initially (this may go well with the new ["swarm testing" mode](https://github.com/googleprojectzero/fuzzilli/commit/9d6145cd4c34b316b6b87e1c5313e4f2c4a0be64))
- For any crashing sample, add a comment to the footer that contains the process arguments as string

If a crash only reproduces with the custom flags, it'll be marked as flaky by the parent. But maybe that's fine? The crash message will state what kind of crash it was (e.g. the debug assertion failure message and stack trace), and that it used custom flags, so maybe that's good enough (unless we expects hundreds of new crashes). Alternatively, it probably wouldn't be too much effort to spawn a new REPRL instance with the custom command line options for reproing crashes, but I'm not convinced it's worth it, at least not initially.

When sharing interesting samples between instances, it's not clear to me what the optimal behaviour would be anyway. If we include the original flags, and from that point on only execute any derived sample with the same flags, then we'll end up constantly restarting the REPRL instance for pretty much every mutation. So we probably don't want that. Also I guess executions that used to be interesting in the past may no longer be relevant once we change flags (e.g. JIT compilation thresholds)? For that reason, I think it's probably easiest if every instance just has a fixed set of random flags that it keeps throughout its lifetime. With [distributed fuzzing](https://github.com/googleprojectzero/fuzzilli/tree/main/Cloud/GCE), we may anyway restart worker instances ever 24h (if they are preemtible instances), so then we'd automatically reshuffle the engine arguments. WDYT?

**Comment by bernhl** [comment]
Your observations are correct; the implementation with the least complexity just adds the command line arguments as comment into the sample (which the PR is doing already, even though the code for placing the comments could be placed better).
Going this route would render reproducing some of the crashes on master instances harder but since they're kept as flaky this is probably fine. If this turns out too much of an issue in practice we can extend the protobuf format down the line.
To summarize, I'll change the code such that:

- `--argumentRandomization=` is either `none` or `initial`
- the arguments are added as comment into the samples
- update the PR so you can have a look

Does this work for you?

**Comment by saelo** [comment]
That sounds great, thanks a lot! So that means we won't need any protobuf changes for this feature then in the end?
Also if --argumentRandomization is now either `none` or `initial`, I guess we could instead just make it a boolean flag `--randomizeEngineArguments` or so? But I'm fine either way!

File: Sources/Fuzzilli/Core/FuzzEngine.swift [hunk 1]
```diff
@@ -27,6 +27,7 @@ extension FuzzEngine {
 
         switch execution.outcome {
             case .crashed(let termsig):
+                program.comments.add("Program crash with options: \(fuzzer.runner.processArguments)", at: .footer)
```
[the remark this answers is not in the document]
Comment (bernhl) [reply Sources/Fuzzilli/Core/FuzzEngine.swift:30]: The minimizer was changed such that the comments are kept during minimization

**Comment by bernhl** [comment]
I just pushed a simpler version of the changes

File: Sources/Fuzzilli/Core/FuzzEngine.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (saelo) [reply Sources/Fuzzilli/Core/FuzzEngine.swift:30]: And I think it actually makes sense in general (even when not doing argument randomization) to include the process arguments in the crashing sample as comment, since they are always needed for reproing crashes

**Comment by saelo** [comment]
Perfect, yeah that looks good and should be easy to maintain. Thanks a lot!
